### ISSUE TITLE
Dolby Digital/DTS passthrough-decoding

### ISSUE THREAD

**Issue by mirhl** [PCSX2/pcsx2#432]
I have been lately [discussing](http://forums.pcsx2.net/Thread-Best-Way-to-Record-Audio-Off-of-PS2) about PCSX2 and its multichannel audio support. 

It seems it's not produced in realtime in-game (besides perhaps some experiments with 4.0 DTS).. though some FMVs used to be encoded with 5.1

Now, I know pro logic is supported, but that's not *real* surround. I read rama once [said](http://forums.pcsx2.net/Thread-SPU2-X-v2?pid=191899#pid191899) we were waiting for proper decoders. 
Then I read gigaherz [claiming](https://github.com/PCSX2/pcsx2/pull/44#issuecomment-44514756) we cannot license any encoders... 
But.. wtf? Why would we need to re-encode audio?

We would just need passthrough (just in case the user was indeed using S/PDIF) and decoding, for everybody else
Afaik FOSS decoders already exist for both standards. 

Any idea?

**Comment by ramapcsx2** [comment]
I'd say it will be difficult to pass the proper bitstream to decoders (hardware more so than software).
The source (PCSX2) will never deliver a steady stream so some kind of buffering will be required.
If that's given at the decoder side, then it could be done, I guess.
Such a decoder would need to have a compatible license and someone needs to include and attach it to the SPU2-X output. SPU2-X itself would need additions to properly detect a bitstream and pass it to the decoder.

**Comment by mirhl** [comment]
I quickly found [liba52](http://liba52.sourceforge.net/) for DD and [libdca](http://www.videolan.org/developers/libdca.html) for DTS. 
Ideally then, [AC3Filter/Spdifer](http://www.ac3filter.net/wiki/Main_Page) should support both (including passthrough). I'm not so apt to really know which could be more or less suitable for PCSX2's needs though. 
They are all licensed with GPL. So green light on that :)

Meanwhile, I even found a nice explanation about the process -including hardware offloading- [under linux](http://www.johannes-bauer.com/linux/dolby/) (I'm sure gregory38 will like this)

EDIT: [ffmpeg](http://www.ffmpeg.org/ffmpeg-codecs.html#ac3) also supports AC3

**Comment by gregory38** [comment]
Be aware that GPL doesn't protect from patents. Liba52 and libdca seems fine but I don't know for ffmpeg.

I would personally prefer a nice and working implementation ;) Sound is already erratic enough on linux unfortunately...

**Comment by mirhl** [comment]
I just posted as much as open-source stuff I could find. Liba52 and libdca would be already enough for pcsx2. Ffmpeg is just another option (we don't need dts-hd or dolby truehd after all). 

Said this, I barely know the difference between pulseaudio and alsa.. so do what you believe it's better

**Comment by mirh** [comment]
Dumping old IRC logs because reasons

> pseudonym: (Example of something we genuinely don't understand: the digital effects engine in the SPU2)
(What we have for that is the result of testing on the psx's spu by one guy... the thing is......... all our testing indicates his model is very very wrong)

> MrKB: Does it cause any issues in PCSX2 though?

> pseudonym: Yes, it makes every game that uses digital effects (which is every game with 3D sound of some sort) sound wrong
You mentioned FFX for example.  Have you listened to the underwater portions on pcsx2 and on a real PS2? :P
Any kind of "room effect" is fucked :P
In addition to sounding wrong, there is actually a chance that it completely breaks some game
Because technically... I don't know if any game does this but it is technically a thing you could do...
..........you could set the SPU2 to interrupt the IOP when a certain address in the effects working area is touched
And rely on that happening at a predictable time
Or at least at a predictable interval
The HD remaster could easily be wrong too
And even if it's not it could easily be different

> MrKB: Could I play the PS2 version on a truly BC PS3? Or does that PS3 still emulate the SPU2?

>pseudonym: The PS3 doesn't use any variant of the SPU so the entire sound engine will be either replaced or emulated for the PS3
I dunno about hw bc ps3s
In any case we absolutely know ours is fucked because of spu2 tests we wrote
We wrote tests which record the waveform produced by playing a sample with an effect applied
Our output was only similar in the broadest of senses

> MrKB: How would one go about understanding it? Reverse engineering?

> pseudonym: Yeah, just got to write tests, make a model that fits, test that model etc.
Until you're 99% sure your model's right
I'm pretty damn sure the guy who originally did this wasn't right even for the psx spu
The SPU2 is supposedly just two SPUs in serial
So if we work this out I'm pretty sure it'd benefit psx emulators too
SPU2 emulation is probably mostly there other than that.  Maybe the slides are still not right.
Well, the DMA isn't but IOP dma's a mess

> MrKB: From the FAQ - PCSX2 has come a long way since its starting point back in 2001. Current features include: Proper SPU2 emulation featuring Time Scaling and Reverb.
lel
Was it ghz|afk that coded it? Or did he only do the work on SPU2?

> pseudonym: That's a funny statement, since if you emulate the SPU2 without "time scaling", all midi style music will be monotone
On the bright side, reversing that is an entirely self contained task. 
A tedious and frustrating one that few people would want to do but it doesn't require anything but the ability to read the SPU2 manual, write test programs and stare at data for days on end

> rama2: yo :p
reading the backlog, some spu2 talk
nice
pseudonym says the reverb is wrong
well, it's about 90&% right

> pseudonym: About 10%

> rama2: it has to be, because if it was that wrong, we would get insane overdrives and feedbacks

> pseudonym: Echoes last no more than 200% longer than they should and all the waveforms are entirely wrong :P
And all other room effects are just fucked.  Completely :P

> rama2: it would never sound like reverb *at all*

> pseudonym: The processing isn't even remotely correct

> rama2: meh, did you ever even compare yourself pseudonym?

> pseudonym: Yes

> rama2: 4 out of 5 games it sounds identical to me

> pseudonym: I blame your hearing
That said, a lot of games don't use the digital effects at all
So those games sounding the same doesn't mean anything

> rama2: square always messes up with their audio drivers :p
i selected reverb using games, stuff like bof3 and xenosaga, remember?
you can't make them sound as good as they do now with a totally broken fx

> pseudonym: The fact that we don't have massive positive feedback loops doesn't mean that it's even remotely correct
It just means that it's completely wrong without being offensive to your ears

> rama2: it's a massive analogue style feedback loop. it breaks very easily and we would notice right aways
since it doesn't do that and instead in most games sounds exactly right (to me! :p ), it has to be pretty close

> pseudonym: not really.

> rama2: okay, which game and which sound in it is broken?

> pseudonym: As I mentioned the obvious test is FFX battles underwater

> rama2: leave that out please, because of other variables
square messes with the spu2 somehow, you know

> pseudonym: Their game magically rewires the SPU2 :P
No, it's just using the hardware features :P

> rama2: we don't get other sounds in the game right either and it could be just that
the problem is their fast updates in some special sounds
psx emu guys wrote about it

> pseudonym: Our _own_ tests showed that the reverb is wrong

> rama2: it's known to be super hard to emulate right
yea, we had a bit lower amplitude and length
so it's like, we lost some phase alignment after the 5th feedback or smth
minimal!°

> pseudonym: Uh and you know... square waves
Everywhere.

> rama2: don't remember those :p
rama2: do you still have screens btw?

> pseudonym: No of course not

> rama2: would be great now :p
lemme look

> pseudonym: Rather than say "it must be anything but the thing we absolutely know is fucked" how about "it's probably the thing we know is fucked"?

> rama2: what, for FFX?
> pseudonym: This is the one thing in SPU2-X nobody has ever seriously tried to fix
> rama2: for ffx i'm saying there are more problems than just reverb :)
> pseudonym: (Because doing so would require reversing it from scratch)
> rama2: yea, we could prolly do it
we did it half assed before

> pseudonym: We didn't really do it at all
Every single SPU or SPU2 emulator is based on the same guy's reversing

> rama2: but it produced tools and methods to do it, and that's the annoying part
so that helps :p
we have a method to load waves onto the ps2 and get the sound back + a waveform onto memory card

> pseudonym: I think for reversing we'll need better than that
The end result is too far removed from the process

> rama2: when redoing it now, i'd start by double checking if that code is all sane though :p

> pseudonym: Watching the effects area during processing is probably more helpful

> rama2: think you can sample and display it?

> pseudonym: I think it's doable.  Not every cycle though
Although repeated testing could get every cycle

> rama2: well, if you had that tool and suppose it does a live display while you upload sounds and program the fx
where do you start reversing the whole thing then?
wasn't it like 30+ registers? :p

> pseudonym: Fuck knows.
Probably start by zeroing the lot and poking them until something happens then think about what just happened a lot :P

> rama2: also are we sure with the memory region at all?

> pseudonym: Yes, the effects area is at least a documented thing

> rama2: yea well
there were issues.. (i forgot what exactly >< )

> pseudonym: Although for all I know it's possible to overflow it if you set the registers to something screwy

> rama2: something with bootup, one game (formula one 2000) and software not doing what the specs say

> pseudonym: The manuals are completely silent about how it actually works
They say "use these presets, don't think too hard about it"

> rama2: yea, that's written for the audio guy :p
he needs to know where the fx on / off and the settings are and sometimes he might care about how much ram it will use
but not more
 i think only that sakuraba motoi guy ever did his own effects :p
and he's crazy. ask the dolphin guys, they have his crap to deal with, too :p
(tales of something )
hmmm
i have recordings but no pics
stuff is from 2010
https://i.imgur.com/xLIiWS9.png
good days :p

> pseudonym: In any case regardless of how "good" you think it is, there's no reason it shouldn't be bit exact

> rama2: the reason is you need the exact specs for everything :)
might be simpler to go and ask sony, lol

> pseudonym: Sony don't know

> rama2: hm, isn't their spu emulation correct?
i think i never checked that
but i agree with you that fixing the fx would be a good start fixing the remaining spu2 issues
only that it's so damn hard :p

> pseudonym: It might be the only remaining spu2 issue apart from dma stuff for all we know :P

> rama2: there's interrupts in the reverb region and if games time that, then fuck

> pseudonym: I really think the voice emulation is... probably right now?
We weren't able to write a test which got voice emulation wrong

> rama2: no, there's this problem with fast updates
it was something about writing to a register really quickly and that would do some effect
FFX
i'm saying that this means our emulation has a problem with resolution

> pseudonym: I prefer to focus on things that are definitely wrong rather than things that are maybe wrong

> rama2: yea but 30 regs and a couple uncerten things.. enjoy :p
thinking about it, it might be fun though
lol
it depends on the tools again

[the discussion than proceeds with pseudonym complaining about gsdx upscaling hacks ^^]
[until I link this very thread and we start talking about surround itself]

> pseudonym: I don't know a lot about surround sound hacks and don't care about them
When I say hacks I mean that surround data encoding in stereo is a hack
Yes I know these are real standards
Stereo for life
(I'm very much the wrong person to ask about surround)

> me: wait wait wiat
this is not "surround encoded in stereo" (aka matrixing)

> pseudonym: The PS2 only does stereo
So yes it is

> me: this is **surround encoded to fit a cable with only the bandwith for two uncompressed channels**
 but compressed surround is still surround

> pseudonym: And I don't really care how you decode it

[we then discover skid_au allegedly has a "newer PS2" with no toslink output, therefore preventing him from doing what I recommend in the first link of this issue]

> MrKB: would [recording digital audio coming from the PS2] potentially make it easier to work with [SPU2 problems in general]?

> pseudonym: Not really.
As I said reversing is probably mostly going to have to be based on the contents of the effects working area, not the final output
We can get output no problem.
SPU2 is in fact designed to be able to download the output to IOP memory
There are two memory areas for mixed audio output
One dry (no effects), the other wet (with effects)
You can simply download from these to see what the output is
That said this isn't the final final output

> me: :p

> pseudonym: There's a final mix of streamed audio, external inputs, dry mixed voices and wet mixed voices to the output
But that's simple enough.
(Note that external input is a technicality in reality)
(Although the SPU2 could theoretically be wired to have an external input, in reality it doesn't, so this is actually used for wiring the two cores of the SPU2 together)
Sorry I said something wrong about dry/wet.  That was stupid.

> gigaherz: technically, the spu2 core0 would still have its own external input, it's just either not plugged to anything, or pluggedto the cdrom for ps1 emulation reasons -- I was never certain (not that I looked too hard)

> pseudonym: At the very least sony tell you to turn off mixing it
 For PS2 software
Plugged into the CDROM sounds plausible

**Comment by MrCK1** [comment]
Using ```libdca``` may violate DTS patents.

Found on VideoLan's website:

```"Provisional Warning: DTS Inc. claims that use of libdca software, to decode DTS compressed sound data on a DVD could violate DTS's patent rights. If you are unsure about the legality of using and distributing this code in your country, in particular in the USA, please consult your lawyer before downloading it."```

**Comment by mirh** [comment]
Ever heard of patent trolling?
It's no different than the mp3, or libdvdcss situation or h264 situation. 
And even DMCA safeguards interoperability (and we aren't talking about encoding). 

Said this, it [seems](https://github.com/foo86/dcadec/issues/15#issuecomment-86236628) libdca is quite outdated compared to libdcadec (for as much as PS2 emu shouldn't need anything past core DTS). 
This in turns looks now totally merged in ffmpeg. 

So.. if just @gigaherz had time :p

**Comment by gigaherz** [comment]
Doesn't matter -- I tried once, and failed epicly. My attempt not only failed to decode any audio, but it prevented all the dolby-enabled games I had from booting. I simply could not figure out the right bit rate that the spdif passthrough works at, or at least my attempts at doing so caused games to stop working. And either way, the data I managed to dump, was never accepted by the lib.

Also, I do have time, I just don't have the brainpower. working a full-time programming job leaves me braindead for the rest of the day. The only moments I have when I can think, are the weekends, and I don't really want to spend those in something like dolby. ;P

**Comment by mirh** [comment]
Well, given you now have reference implementations to compare, it should be quite easier (prolly still not *easy*). 

Said this, we know the story, it's all in the logs I posted on the forums :)

**Comment by gigaherz** [comment]
Reference implementations? Where? Did I miss something?

**Comment by mirh** [comment]
Not in the sense "specific to playstation 2", but as I said AC3 and DTS are pretty thoroughly documented standards. 
And I believe it's far easier to make stuff work, when you don't have to decrypt an audio compression algorithm too, among other things.

**Comment by gigaherz** [comment]
Yeah I never attempted to -- I did use liba52 in my attempt, I just failed at being able to provide liba52 with any sort of meaningful bitstream. Maybe things have changed, but I'm not really interested in trying again. ;P

**Comment by kenshen112** [comment]
https://github.com/kodi-adsp/adsp.freesurround

An open source implementation that seems to be useable outside of Kodi?

**Comment by Filoppi** [comment]
For your information, I'm reworking DPLII on Dolphin, and it's not been easy (I made some changes and fixes to FreeSurround).
https://github.com/Filoppi/dolphin/blob/audio-input-fixes/Externals/FreeSurround/include/FreeSurround/FreeSurroundDecoder.h
The dolphin decoder was taken from [https://hydrogenaud.io/index.php?topic=52235.0](url) but adsp.freesurround is based on the same source, though it's probably cleaner and better.

**Comment by ThreeDeeJay** [comment]
I second FreeSurround, hands down
It's [miles better than any alternative I've tested](https://airtable.com/shrWKD1FSh3oOYySP), even compared to the PCSX2 decoder setup (`Audio Expansion Mode=Surround` + `SPU2-X.ini [OUTPUT] DplDecodingLevel=0`).
In the meantime, there's [EqualizerAPO/HeSuVi decoders](https://github.com/Dogway/emulation-random/tree/master/EqualizerAPO/Surround) that do an okay job, though. Here's some footage I recorded using them.

[![IMAGE ALT TEXT](https://img.youtube.com/vi/XN_bhP7AyD8/hqdefault.jpg)](https://youtu.be/XN_bhP7AyD8 "Need For Speed: Most Wanted [PCSX2] | HeSuVi + Dolby Pro Logic II | Headphone Virtual Surround 5.1")

[![IMAGE ALT TEXT](https://img.youtube.com/vi/-Yu8saszVpY/hqdefault.jpg)](https://youtu.be/-Yu8saszVpY "Need For Speed: Most Wanted [PCSX2] | HeSuVi + Dolby Pro Logic II | Headphone Virtual Surround 5.1")

**Comment by ThreeDeeJay** [comment]
It's been a while, and I know there's not a lot activity here, but since it's relevant, here's another, recent test: https://youtu.be/a5w-DzR0Xlg
[![IMAGE ALT TEXT](https://img.youtube.com/vi/a5w-DzR0Xlg/hqdefault.jpg)](https://youtu.be/a5w-DzR0Xlg "Star Ocean: Till the End of Time [PCSX2] | Dolby Pro-Logic II (Virtual +) 5.1 Surround decoding")

**Comment by ISuckGood** [comment]
I have dolby surround in games in pcsx2 and rpcs3 . I just select dolby/dts in my sound card and connected the speakers to my sound card via optical . I select 5.1 in the emus . Voice comes from center channel only and rear speakers have proper sound. I tested it in jak and daxter on pssx2 and rpsc3

**Comment by ThreeDeeJay** [comment]
@ISuckGood I think PS3 games use Dolby Digital/DTS, which output discrete 5.1 surround channels, as opposed to Dolby Pro-Logic II in PS2, where surround channels have to be extracted from a stereo signal.
I read some PS2 games use Dolby Digital, but apparently PCSX2 doesn't support it at all.

**Comment by MMatiasMG** [comment]
Any news?

**Comment by F0bes** [comment]
No

**Comment by stenzek** [comment]
@MMatiasMG please stop bumping issues, if there is anything to be seen, it's clearly visible.

**Comment by seta-san** [comment]
For those worried about patents. Patents only last 20 years.

**Comment by stenzek** [comment]
We are aware. I've brought the AC3/DTS patent situation up internally.

AFAIK it's still unknown how the bitstream is handled by the SPU. Someone would need to reverse that.

**Comment by Zoomer1988** [comment]
But the Dolby Prologic II decoding has nothing to do with the bitstream, right? Have you guys also been discussing using the aforementioned FreeSurround implementation in PCSX2?

**Comment by Ziemas** [comment]
> But the Dolby Prologic II decoding has nothing to do with the bitstream, right?

Right, the freesurround discussion was off topic to this issue in the first place, this issue is about decoding the digital formats.

**Comment by Zoomer1988** [comment]
Yeah, fair enough. I've re-read the issue, and it's "Not a real surround" apparently.

It seems, though, that this is the only place, where DPLII is being properly discussed at all with examples and the FreeSurround foobar DSP (the Dolphin implementation) is even mentioned.

Still, has nothing to do with digital stream decoding indeed. My bad, sorry.

**Comment by ThreeDeeJay** [comment]
Yeah, usually [DPL decoding discussions were redirected here](https://github.com/PCSX2/pcsx2/issues/3218#issuecomment-595995071).
And in defense of DPL, while I agree matrix-encoded surround in stereo is inferior to raw/discrete channels because of inherent quality loss, I don't think it's fair or even accurate to say it's "not real surround" given it's [decoded properly](https://airtable.com/appayGNkn3nSuXkaz/shrWKD1FSh3oOYySP/tblZ8QmkQmfZuoquV/viw0mS6nvHZ2x2gUx/recoP3vrx7PvKdBJy), like it does now after [this update](https://github.com/PCSX2/pcsx2/pull/11135): https://youtu.be/fWIUJqPfpNA
(Just a heads up for anyone finding/watching this thread for DPL decoding. I'll refrain from discussing this here any further to let discussion focus on passthrough)
